pkg/webhook/resources/virtualmachine/validator.go

```diff
@@ -69,12 +75,134 @@ func (v *vmValidator) Resource() types.Resource {
+	if err != nil {
+		return clusterNetwork, err
+	}
+	clusterNetwork, ok := nad.Labels[keyClusterNetwork]
+	if !ok {
+		return clusterNetwork, err
+	}
+
+	return clusterNetwork, nil
+}
+
+func (v *vmValidator) getVMInterfaceInfo(vms []*kubevirtv1.VirtualMachine) (MacAddrPerCluster map[string]map[string]string, err error) {
+	MacAddrPerCluster = make(map[string]map[string]string) //map[clusterName]map[macaddr]intfName
+
+	for _, vm := range vms {
+		vmName := vm.Namespace + "/" + vm.Name
+
+		if vm.Spec.Template == nil {
+			return MacAddrPerCluster, fmt.Errorf("vm %s template is nil", vm.Name)
+		}
+
+		vmCluster := make(map[string]string) // map[netIntfName]clusterNetwork
+		vmNetworks := vm.Spec.Template.Spec.Networks
+		for _, vmNetwork := range vmNetworks {
+			if vmNetwork.Multus == nil || vmNetwork.Multus.NetworkName == "" {
+				continue
+			}
+			cn, err := v.getClusterNetworkForNad(vmNetwork.Multus.NetworkName)
+			if err != nil {
+				return MacAddrPerCluster, err
+			}
+			vmCluster[vmNetwork.Name] = cn
+		}
+
+		vmInterfaces := vm.Spec.Template.Spec.Domain.Devices.Interfaces
+
+		var macAddrs map[string]string
+		for _, vmInterface := range vmInterfaces {
+			clusNet := vmCluster[vmInterface.Name]
+			if _, exists := MacAddrPerCluster[clusNet]; !exists {
+				macAddrs = make(map[string]string)
+			} else {
+				macAddrs = MacAddrPerCluster[clusNet]
+			}
+			macAddrs[vmInterface.MacAddress] = vmInterface.Name + "+" + vmName
+			MacAddrPerCluster[clusNet] = macAddrs
+		}
+	}
+
+	return MacAddrPerCluster, nil
+}
+
+func (v *vmValidator) checkForDuplicateMacAddrs(vm *kubevirtv1.VirtualMachine) (err error) {
+	var newvm []*kubevirtv1.VirtualMachine
+	var oldvmsInfo map[string]map[string]string
+	var oldvmsList []*kubevirtv1.VirtualMachine
+
+	//get map[clustername]map[mac]string info for new vm
+	newvm = append(newvm, vm)
+	newVMInfo, err := v.getVMInterfaceInfo(newvm)
+	if err != nil {
+		return err
+	}
+
+	if vm.Spec.Template == nil {
+		return fmt.Errorf("vm %s template is nil", vm.Name)
+	}
+
+	vmInterfaces := vm.Spec.Template.Spec.Domain.Devices.Interfaces
+
+	for _, vmInterface := range vmInterfaces {
+		oldvms, err := v.vmCache.GetByIndex(indexerwebhook.VMByMacAddress, vmInterface.MacAddress)
+		if err != nil {
+			return err
+		}
+
+		oldvmsList = append(oldvmsList, oldvms...)
+	}
+
+	oldvmsInfo, err = v.getVMInterfaceInfo(oldvmsList)
+	if err != nil {
+		return err
+	}
+
+	for cn, macs := range newVMInfo {
+		if othermacs, exists := oldvmsInfo[cn]; exists {
+			if comparemacs(macs, othermacs) { //duplicate present
+				return fmt.Errorf("duplicate mac address present for vm %s for cluster network %s", vm.Name, cn)
+			}
+		}
+	}
+
+	return nil
+}
+
+func comparemacs(macList1 map[string]string, macList2 map[string]string) bool {
+	for mac1, newvmIntfName := range macList1 {
+		if oldvmIntfName, exists := macList2[mac1]; exists && oldvmIntfName != newvmIntfName { //skip same vms during vm migration,restore cases
```

Comment:
seems if `oldvmIntfName == newvmIntfName` indicates that the vms are from migration/restore cases,how is this ensured?

suppose use one vm's yaml to create another vm, and the only change is vmname, will above check identify it? thanks.

Reply:
yeah oldvmIntfName/newvmIntfName is a combination of both interface name and vm name.